### PR TITLE
Add Connector for Confluent Schema Registry 

### PR DESCRIPTION
**Pull request by nharring-adjacent**
This PR continues the work of adding Connectors, adding support for creating, dropping, and using connectors for Confluent Schema Registry. To that end `CsrConnectorAvro` and `CsrConnectorProto` were modified to contain a new `CsrConnector` enum with `Reference` and `Inline` variants with the same semantics as the `KafkaConnector` enum variants. The parser was updated to accept new syntax, specifically `FORMAT AVRO CONFLUENT SCHEMA REGISTRY CONNECTOR <connector_name>`.

This change required rethinking how to handle connectors and purification which allow for reverting the previous change to `purify_create_source` which removed the `async` keyword in favor of returning an `impl Future` from a nested `async move` block. To accomplish this the `Reference` variants of each connector got optional fields which are populated with None by the parser and then reified before purification or planning (for new creation vs. reading from catalog) to contain `Some(T)` based on the contents of the connector in question. I took this opportunity to also revisit how name expansion was done for connectors with a similar approach to prevent a big explosion in complexity inside `normalize::create_source_statement` in the `CreateSourceStatement` match arm.

### Motivation

  * This PR adds a known-desirable feature.

   [Connnectors](https://github.com/MaterializeInc/materialize/issues/11504

 
### Tips for reviewer
This PR is split into 3 commits which are fairly well grouped around the changes contained in them, though none of them are self-contained enough to merge on their own.

### Testing
I have added a number of test cases which I believe cover both the core functionality and some important edge cases.

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - This PR extends the Connector feature to allow for creating and using connectors for `CONFLUENT SCHEMA REGISTRY` properties.

### CODE REVIEW

File: src/sql/src/plan/statement/ddl.rs [hunk 1]
```diff
@@ -1390,7 +1423,10 @@ fn get_encoding_inner(
                     seed
                 {
                     let mut ccsr_with_options = normalize::options(&ccsr_options);
-
+                    let url = match connector {
```
Comment (sploiselle) [inline src/sql/src/plan/statement/ddl.rs:1426]: Returning an empty string as a url to be parsed for the ccsr config doesn't seem quite right to me. Does this only want to occur if you're using an inline definition because you do this more meaningfully in `purify_csr_connector_proto`?

```diff
diff --git a/src/sql/src/plan/statement/ddl.rs b/src/sql/src/plan/statement/ddl.rs
index b62d3f53b..782a187bb 100644
--- a/src/sql/src/plan/statement/ddl.rs
+++ b/src/sql/src/plan/statement/ddl.rs
@@ -1423,16 +1423,14 @@ fn get_encoding_inner(
                     seed
                 {
                     let mut ccsr_with_options = normalize::options(&ccsr_options);
-                    let url = match connector {
-                        mz_sql_parser::ast::CsrConnector::Inline { url: uri } => uri,
-                        mz_sql_parser::ast::CsrConnector::Reference { .. } => "",
-                    };
-                    // We validate here instead of in purification, to match the behavior of avro
-                    let _ccsr_config = kafka_util::generate_ccsr_client_config(
-                        url.parse()?,
-                        &kafka_util::extract_config(&mut normalize::options(with_options))?,
-                        &mut ccsr_with_options,
-                    )?;
+                    if let mz_sql_parser::ast::CsrConnector::Inline { url } = connector {
+                        // We validate here instead of in purification, to match the behavior of avro
+                        let _ccsr_config = kafka_util::generate_ccsr_client_config(
+                            url.parse()?,
+                            &kafka_util::extract_config(&mut normalize::options(with_options))?,
+                            &mut ccsr_with_options,
+                        )?;
+                    }
```

File: src/sql/src/plan/statement/ddl.rs [hunk 2]
```diff
@@ -1903,7 +1939,7 @@ fn kafka_sink_builder(
         Some(Format::Avro(AvroSchema::Csr {
             csr_connector:
                 CsrConnectorAvro {
-                    url,
+                    connector: CsrConnector::Inline { url: uri },
```
Comment (sploiselle) [inline src/sql/src/plan/statement/ddl.rs:1942]: nit: renaming this `uri` and then calling `uri.parse::<Url>()?` feels extra persnickety; I'd just call it a url

File: src/sql/src/pure.rs [hunk 3]
```diff
@@ -382,13 +363,22 @@ async fn purify_csr_connector_proto(
...
         with_options: ccsr_options,
     } = csr_connector;
     match seed {
         None => {
-            let url: Url = url.parse()?;
+            let url: Url = match connector {
+                CsrConnector::Inline { url: uri } => uri.parse()?,
+                CsrConnector::Reference { url, .. } => match url {
+                    Some(url) => {
+                        info!("parsing {} into url", url);
```
Comment (sploiselle) [inline src/sql/src/pure.rs:376]: This kind of thing is great for debugging, but to be useful in system logging, you'd likely want to include more context for some random DBA/sales engineer as to why this could point to a problem in the first place--i.e. put yourself in the user's shoes

File: src/sql/src/pure.rs [hunk 4]
```diff
@@ -434,12 +424,18 @@ async fn purify_csr_connector_avro(
...
-        url,
+        connector,
         seed,
         with_options: ccsr_options,
     } = csr_connector;
     if seed.is_none() {
-        let url = url.parse()?;
+        let url = match connector {
+            CsrConnector::Inline { url } => url.parse()?,
+            CsrConnector::Reference { url, .. } => match url {
```
Comment (sploiselle) [inline src/sql/src/pure.rs:434]: think this should be `expect`

File: src/sql/src/pure.rs [hunk 5]
```diff
@@ -382,13 +363,22 @@ async fn purify_csr_connector_proto(
...
+        connector,
         seed,
         with_options: ccsr_options,
     } = csr_connector;
     match seed {
         None => {
-            let url: Url = url.parse()?;
+            let url: Url = match connector {
+                CsrConnector::Inline { url: uri } => uri.parse()?,
+                CsrConnector::Reference { url, .. } => match url {
```
Comment (sploiselle) [inline src/sql/src/pure.rs:374]: think this should be `expect`

File: src/coord/src/catalog.rs [hunk 6]
```diff
@@ -3324,12 +3324,14 @@ impl mz_sql::catalog::CatalogConnector for Connector {
     fn uri(&self) -> String {
```
[the remark this answers is not in the document]
Comment (nharring-adjacent) [reply src/coord/src/catalog.rs:3324]: I actually forgot I implemented these on `ConnectorInner` so I can just delegate.

File: src/sql-parser/src/ast/defs/ddl.rs [hunk 7]
```diff
@@ -120,23 +120,47 @@ impl<T: AstInfo> AstDisplay for ProtobufSchema<T> {
...
+    Inline {
+        url: String,
+    },
+    // The reference variant can have url and with_options populated for use in purification
+    // or any other context where
+    Reference {
+        connector: UnresolvedObjectName,
+        url: Option<String>,
+        // Vec of keys interspersed with values
+        with_options: Option<Vec<String>>,
```
[the remark this answers is not in the document]
Comment (nharring-adjacent) [reply src/sql-parser/src/ast/defs/ddl.rs:134]: `walkabout` cannot currently handle `BTreeMap` or `Vec<(String, String)>` and when I mentioned it in slack we agreed to start with the version that required the least code.

File: src/sql/src/connectors.rs [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+
+use crate::normalize::unresolve;
+use crate::normalize::unresolved_object_name;
+use crate::plan::PlanError;
+use crate::plan::StatementContext;
+use crate::{catalog::SessionCatalog, normalize};
+
+/// Uses the provided catalog to populate all Connector references with the values of the connector
+/// it references allowing it to be used as if there was no indirection
+pub fn reify_connectors<T: AstInfo>(
```
[the remark this answers is not in the document]
Comment (nharring-adjacent) [reply src/sql/src/connectors.rs:25]: I agonized over naming this but felt that `resolve` was already heavily used and wanted to avoid confusion and thought that this does actually specify pretty clearly what's happening by taking an abstract reference and transforming it to concrete values.

That said I am very loosely attached to this specific term and will gladly adopt something that will be more effective. Perhaps `populate` or `hydrate`?

File: src/sql/src/connectors.rs [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+// by the Apache License, Version 2.0.
+
+use mz_repr::GlobalId;
+use mz_sql_parser::ast::{
+    AstInfo, AvroSchema, CreateSourceConnector, CreateSourceFormat, CreateSourceStatement,
+    CsrConnector, CsrConnectorAvro, CsrConnectorProto, Format, KafkaConnector,
+    KafkaSourceConnector, ProtobufSchema, UnresolvedObjectName,
+};
+
+use crate::normalize::unresolve;
```
[the remark this answers is not in the document]
Comment (nharring-adjacent) [reply src/sql/src/connectors.rs:17]: I had thought `cargo fmt` handled getting the imports right since I let automation do all of the adding and arranging of those. I also thought I had de-qualified everything that was imported but apparently I missed some once it stopped triggering the unused import rules. Thanks for catching this!

File: src/sql/src/connectors.rs [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+            qualify_connector_in_single_schema(key, scx)?;
+            qualify_connector_in_single_schema(value, scx)?;
+        }
+    }
+
+    Ok(())
+}
+
+/// Helper function to resolve names for connectors in a single [`Format`]
+fn qualify_connector_in_single_schema<T: AstInfo>(
```
[the remark this answers is not in the document]
Comment (nharring-adjacent) [reply src/sql/src/connectors.rs:181]: This should be `format`, the inner value in each format variant is a type with `schema` in its name but its usage here was a mistake.

File: src/sql/src/plan/statement/ddl.rs [hunk 11]
```diff
@@ -355,18 +358,29 @@ pub fn plan_create_source(
...
+                    broker,
+                    with_options,
+                    connector,
+                } => {
+                    let broker_addr = match broker {
+                        Some(url) => url.to_owned(),
+                        None => bail!("resolved connector must specify broker address"),
+                    };
+                    let options = match with_options {
+                        Some(opts) => BTreeMap::from_iter(opts.iter().cloned().tuples()),
```
[the remark this answers is not in the document]
Comment (nharring-adjacent) [reply src/sql/src/plan/statement/ddl.rs:375]: I will probably need some guidance on how to write the proper addition to `walkabout` to enable that, while I think I mostly get what that crate does the details of the implementation are still opaque to my brain.

File: src/sql/src/plan/statement/ddl.rs (hunk 1 above)
Comment (nharring-adjacent) [reply src/sql/src/plan/statement/ddl.rs:1426]: Oh dammit, this should not have slipped through. I am actually completely stunned that this even passes tests, I had written it this way originally to make the `match` arms type check while I was fleshing out other details and all of the other places I took a similar shortcut and forgot to fix emerged in testing quite rapidly. 

I will fix this but I also need to figure out why it did not get caught.

File: src/sql/src/plan/statement/ddl.rs (hunk 2 above)
Comment (nharring-adjacent) [reply src/sql/src/plan/statement/ddl.rs:1942]: Agreed, this was a transitionary naming thing that I I had caught all of.

File: src/sql/src/pure.rs [hunk 12]
```diff
@@ -55,172 +55,153 @@ use crate::normalize;
...
+            let (broker, connector_options) = match connector {
+                KafkaConnector::Reference {
+                    broker,
+                    with_options,
+                    ..
+                } => {
+                    let broker_url = match broker {
+                        Some(url) => url.to_owned(),
+                        None => {
+                            bail!("Failed to purify source because connector remained unresolved")
```
[the remark this answers is not in the document]
Comment (nharring-adjacent) [reply src/sql/src/pure.rs:89]: This is the code path that someone will encounter if they specify the wrong name for a connector so I think this should be an error back to the user.

File: src/sql/src/pure.rs (hunk 3 above)
Comment (nharring-adjacent) [reply src/sql/src/pure.rs:376]: Yup, this should've been removed. The big lesson for me here is this PR got too big for me to keep track of all of these type of detail and clean it up properly before asking for review. ;)

File: src/sql/src/pure.rs (hunk 5 above)
Comment (nharring-adjacent) [reply src/sql/src/pure.rs:374]: When I tried using expect here I ran into
```
cannot move out of `*url` which is behind a mutable reference
move occurs because `*url` has type `std::option::Option<std::string::String>`, which does not implement the `Copy` trait
```
and adding `as_ref()` as suggested then breaks type inference.

Comment (sploiselle) [reply src/sql/src/pure.rs:371]: Hmm, I'm not able to duplicate that issue at all, and this code works for me w/o issue:

```diff
diff --git a/src/sql/src/pure.rs b/src/sql/src/pure.rs
index e95253637..357b8728e 100644
--- a/src/sql/src/pure.rs
+++ b/src/sql/src/pure.rs
@@ -23,7 +23,6 @@ use mz_sql_parser::ast::{CsrConnector, KafkaConnector, KafkaSourceConnector};
 use prost::Message;
 use protobuf_native::compiler::{SourceTreeDescriptorDatabase, VirtualSourceTree};
 use protobuf_native::MessageLite;
-use reqwest::Url;
 use tokio::fs::File;
 use tokio::io::AsyncBufReadExt;
 use tokio::task;
@@ -366,13 +365,11 @@ async fn purify_csr_connector_proto(
     } = csr_connector;
     match seed {
         None => {
-            let url: Url = match connector {
-                CsrConnector::Inline { url: uri } => uri.parse()?,
-                CsrConnector::Reference { url, .. } => match url {
-                    Some(url) => url.parse()?,
-                    None => bail!("CSR Connector must specify a url!"),
-                },
-            };
+            let url = match connector {
+                CsrConnector::Inline { url } => url,
+                CsrConnector::Reference { url, .. } => url.as_ref().unwrap(),
+            }
+            .parse()?;
             let kafka_options = kafka_util::extract_config(&mut normalize::options(with_options))?;
             let ccsr_config = kafka_util::generate_ccsr_client_config(
                 url,

```

Let me know if you'd like to chat about how `Option`/`unwrap`/`as_ref` interact. Certainly not the most intuitive paradigm to work within.

File: src/sql/src/pure.rs (hunk 4 above)
Comment (sploiselle) [reply src/sql/src/pure.rs:428]: Just a ping here that this should also use the code from the prior comment

File: src/walkabout/src/ir.rs [hunk 13]
```diff
@@ -315,6 +319,29 @@ fn analyze_type(ty: &syn::Type) -> Result<Type> {
                     "Vec" => container(Type::Vec),
                     "Option" => container(Type::Option),
                     "Box" => container(Type::Box),
+                    "BTreeMap" => match &segment.arguments {
+                        syn::PathArguments::None => bail!("Map type missing arguments"),
+                        syn::PathArguments::AngleBracketed(args) if args.args.len() == 2 => {
+                            let key = match args.args.first().unwrap() {
```
Comment (sploiselle) [inline src/walkabout/src/ir.rs:325]: nit: These calls to `first` and `last` feel a little overwrought to me:

```diff
diff --git a/src/walkabout/src/ir.rs b/src/walkabout/src/ir.rs
index 7dea7a3f3..03ba42e9d 100644
--- a/src/walkabout/src/ir.rs
+++ b/src/walkabout/src/ir.rs
@@ -322,11 +322,11 @@ fn analyze_type(ty: &syn::Type) -> Result<Type> {
                     "BTreeMap" => match &segment.arguments {
                         syn::PathArguments::None => bail!("Map type missing arguments"),
                         syn::PathArguments::AngleBracketed(args) if args.args.len() == 2 => {
-                            let key = match args.args.first().unwrap() {
+                            let key = match &args.args[0] {
                                 syn::GenericArgument::Type(t) => t,
                                 _ => bail!("Invalid argument to map container, should be a Type"),
                             };
-                            let value = match args.args.last().unwrap() {
+                            let value = match &args.args[1] {
                                 syn::GenericArgument::Type(t) => t,
                                 _ => bail!("Invalid argument to map container, should be a Type"),
                             };
```

**Review by sploiselle — APPROVED** [review]
Great work on extending walkabout to support `BTreeMap`.

LGTM modulo comments; feel free to ping me if you run into any issues with them.

File: src/sql/src/pure.rs (hunk 5 above)
Comment (nharring-adjacent) [reply src/sql/src/pure.rs:371]: I was failing to notice moving the `parse()?` to the end of the match block which lets me use `unwrap()` or `expect()` as you suggest.
